### DISCUSSION TITLE
How does the compression for `.webp` work? If an image is compressed, it can be re-compressed at infinitum?

### DISCUSSION THREAD

**Discussion by oscard0m** [libvips/libvips#3371] [Q&A]
# Question
I open this discussion to understand better how the compression for `.webp` works. 
When compressing an image, in this case with `.webp` format, I would expect to get a version with its file size decreased (and a certain amount of quality lost as well), but when running a second time with the same parameters, to obtain the exact image size and quality. The compression algorithm ran before and did all it could to compress the image.

Is this how the image compression for `.webp` works?

# Context
I'm using [Calibre App's Image Action](https://github.com/calibreapp/image-actions) (which uses [sharp](https://github.com/lovell/sharp) -> which uses [libvips](https://github.com/libvips/libvips)), and every time the action runs, it gets more compressed `.webp` images (but not for the rest of formats).

According to [this issue](https://github.com/calibreapp/image-actions/issues/173), `cwebp` offers a `-pass` flag to control that, but I think `libvips` does not offer that for `webp`, but I'm not sure.

Could somebody give me more light here?

Thanks

**Comment by lovell** [comment]
> when running a second time with the same parameters, to obtain the exact image size and quality

This is not the case when using lossy formats, and lossy WebP is particularly lossy.

The `pass` option of `cwebp` relates to its entropy analysis calculations for lossless compression.

**Reply by oscard0m** [reply]
> > when running a second time with the same parameters, to obtain the exact image size and quality
> 
> This is not the case when using lossy formats, and lossy WebP is particularly lossy.

Any recommendation on what I could use then? Do you think `lossless` option would grant "second pass" of the algorithm to not change the result? The compromise is less resulting compression as a result, ofc.

> The `pass` option of `cwebp` relates to its entropy analysis calculations for lossless compression.

Do you mean that even if I use the `pass` argument, each iteration will re-compress the result from the previous one?

**Reply by lovell** [reply]
Perhaps try `nearLossless` WebP as a compromise? Alternatively you could modify the source code of the Action to include the concept of a threshold for change in file size e.g. only re-compress if >5% smaller.

The `pass` option is not exposed by libvips, and I believe it only applies to lossless WebP.
